### PR TITLE
Fix for #1556 - Removing support for llvm versions before 6.0.

### PR DESCRIPTION
**Pull request by DeepakRajendrakumaran**

### CODE REVIEW

**Comment by dbabokin** [comment]
Have you checked that binaries are identical with LLVM 6.0+ ?

Also Travis reports that build with 6.0 is broken. Is it reproducible?

**Comment by DeepakRajendrakumaran** [comment]
Compared binaries with 8.0. Build with 8.0 and 9.0. I created a pull request to run it through Travis. I'm looking into the failure.

File: alloy.py [hunk 1]
```diff
@@ -195,9 +165,7 @@ def checkout_LLVM(component, use_git, version_LLVM, revision, target_dir, from_v
 # present in LLVM libraries. In LLVM 5.0 they are not there by default and require explicit enabling.
 # In later version this functionality is triggered by enabling assertions.
 def get_llvm_enable_dump_switch(version_LLVM):
-    if version_LLVM in ["3_2", "3_3", "3_4", "3_5", "3_6", "3_7", "3_8", "3_9", "4_0"]:
-        return ""
-    elif version_LLVM == "5_0":
+    if version_LLVM == "5_0":
```
Comment (dbabokin) [inline alloy.py]: 5.0 is not supported.

File: src/builtins.cpp [hunk 2]
```diff
@@ -915,33 +865,16 @@ void AddBitcodeToModule(const unsigned char *bitcode, int length, llvm::Module *
...
-                            "Module DataLayout is incompatible with "
-                            "library DataLayout:\n"
-                            "Module  DL: %s\n"
-                            "Library DL: %s\n",
-                            module->getDataLayout().c_str(), bcModule->getDataLayout().c_str());
-                }
-#endif
             }
 #endif
                 bcModule->setTargetTriple(mTriple.str());
```
Comment (dbabokin) [inline src/builtins.cpp]: Is ident of this line right?

File: src/builtins.cpp [hunk 3]
```diff
@@ -915,33 +865,16 @@ void AddBitcodeToModule(const unsigned char *bitcode, int length, llvm::Module *
...
-        if (llvm::Linker::LinkModules(module, bcModule, llvm::Linker::DestroySource, &linkError))
-            Error(SourcePos(), "Error linking stdlib bitcode: %s", linkError.c_str());
-#elif ISPC_LLVM_VERSION <= ISPC_LLVM_3_7 // 3.6-3.7
-        llvm::Linker::LinkModules(module, bcModule);
-#else                                    // LLVM 3.8+
-                                         // A hack to move over declaration, which have no definition.
-                                         // New linker is kind of smart and think it knows better what to do, so
-                                         // it removes unused declarations without definitions.
-                                         // This trick should be legal, as both modules use the same LLVMContext.
+        // LLVM 3.8+
```
Comment (dbabokin) [inline src/builtins.cpp]: No need to mention 3.8 anymore.

File: src/builtins.cpp [hunk 4]
```diff
@@ -973,58 +905,28 @@ static void lDefineConstantInt(const char *name, int val, llvm::Module *module,
...
-        Assert(sym_const_storagePtr);
-        llvm::DIGlobalVariable var = m->diBuilder->createGlobalVariable(file, name, name, file, 0 /* line */, diType,
-                                                                        true /* static */, sym_const_storagePtr);
-#elif ISPC_LLVM_VERSION >= ISPC_LLVM_3_7 && ISPC_LLVM_VERSION <= ISPC_LLVM_3_9 // LLVM 3.7 - 3.9
-        llvm::Constant *sym_const_storagePtr = llvm::dyn_cast<llvm::Constant>(sym->storagePtr);
-        Assert(sym_const_storagePtr);
-        m->diBuilder->createGlobalVariable(cu, name, name, file, 0 /* line */, diType, true /* static */,
-                                           sym_const_storagePtr);
-#else                                                                          // LLVM 4.0+
+        // LLVM 4.0+
```
Comment (dbabokin) [inline src/builtins.cpp]: No need to mention 4.0 anymore.

File: src/cbackend.cpp [hunk 5]
```diff
@@ -5517,61 +5111,36 @@ class MaskOpsCleanupPass : public llvm::BasicBlockPass {
...
-        andNotFuncs[0]->addFnAttr(llvm::Attributes::ReadNone);
-#else /* LLVM 3.3+ */
         andNotFuncs[0]->addFnAttr(llvm::Attribute::NoUnwind);
         andNotFuncs[0]->addFnAttr(llvm::Attribute::ReadNone);
-#endif
         andNotFuncs[1] =
-#if ISPC_LLVM_VERSION <= ISPC_LLVM_4_0
-            llvm::dyn_cast<llvm::Function>(m->getOrInsertFunction("__and_not2", mt, mt, mt, NULL));
-#elif ISPC_LLVM_VERSION <= ISPC_LLVM_8_0 // LLVM 5.0-LLVM 8.0
+#if ISPC_LLVM_VERSION <= ISPC_LLVM_8_0 // LLVM 5.0-LLVM 8.0
```
Comment (dbabokin) [inline src/cbackend.cpp]: I think "5.0-8.0" comment is not relevant anymore

File: src/cbackend.cpp [hunk 6]
```diff
@@ -5517,61 +5111,36 @@ class MaskOpsCleanupPass : public llvm::BasicBlockPass {
...
-#else /* LLVM 3.3+ */
         notFunc->addFnAttr(llvm::Attribute::NoUnwind);
         notFunc->addFnAttr(llvm::Attribute::ReadNone);
-#endif
 
         andNotFuncs[0] =
-#if ISPC_LLVM_VERSION <= ISPC_LLVM_4_0
-            llvm::dyn_cast<llvm::Function>(m->getOrInsertFunction("__and_not1", mt, mt, mt, NULL));
-#elif ISPC_LLVM_VERSION <= ISPC_LLVM_8_0 // LLVM 5.0-LLVM 8.0
+#if ISPC_LLVM_VERSION <= ISPC_LLVM_8_0 // LLVM 6.0-LLVM 8.0
```
Comment (dbabokin) [inline src/cbackend.cpp]: I think "6.0-8.0" comment is not relevant anymore

File: src/ctx.cpp [hunk 7]
```diff
@@ -373,26 +322,12 @@ FunctionEmitContext::FunctionEmitContext(Function *func, Symbol *funSym, llvm::F
...
-        diSubprogram = m->diBuilder->createFunction(diFile /* scope */, funSym->name, mangledName, diFile, firstLine,
-                                                    diSubprogramType_n, isStatic, true, /* is defn */
-                                                    firstLine, flags, isOptimized, llvmFunction);
-        AssertPos(currentPos, diSubprogram.Verify());
-#elif ISPC_LLVM_VERSION == ISPC_LLVM_3_7                                       /* LLVM 3.7 */
-        diSubprogram = m->diBuilder->createFunction(diFile /* scope */, funSym->name, mangledName, diFile, firstLine,
-                                                    diSubprogramType_n, isStatic, true, /* is defn */
-                                                    firstLine, flags, isOptimized, llvmFunction);
-#elif ISPC_LLVM_VERSION == ISPC_LLVM_3_8 || ISPC_LLVM_VERSION == ISPC_LLVM_3_9 /* LLVM 3.8, 3.9 */
+#if ISPC_LLVM_VERSION >= ISPC_LLVM_6_0 && ISPC_LLVM_VERSION <= ISPC_LLVM_7_1 /* LLVM 6.0 to 7.1 */
```
Comment (dbabokin) [inline src/ctx.cpp]: Looks like it can be rewritten as ISPC_LLVM_VERSION < ISPC_LLVM_8_0

File: src/ispc.h [hunk 8]
```diff
@@ -41,7 +41,7 @@
 #include "ispc_version.h"
 
 #if ISPC_LLVM_VERSION < OLDEST_SUPPORTED_LLVM || ISPC_LLVM_VERSION > LATEST_SUPPORTED_LLVM
-#error "Only LLVM 3.2 - 8.0 and 9.0 development branch are supported"
+#error "Only LLVM 6.0 - 8.0 and 9.0 development branch are supported"
```
Comment (dbabokin) [inline src/ispc.h]: 6.0-9.0 and 10.0 development branch

File: src/cbackend.cpp (hunk 6 above)
Comment (dbabokin) [reply src/cbackend.cpp]: Yes, it's 6.0-8.0, given that we support only 6.0+, but my point is that this comment is not informative anymore and probably it's better to remove. I.e. it's a binary choice now between "before 8.0" and "8.0+".

File: src/builtins.cpp (hunk 2 above)
Comment (dbabokin) [reply src/builtins.cpp]: ok

**Pull request merged**